# Do an alignment for each phylum with all seqs? See what fits best? It will be easy in a snakefile
PHYLA, = glob_wildcards('../resources/RdRp-scan/Phylogenies/{ident}.CLUSTALO_0.4.fa')

rule all:
    input:
        expand("../results/alignments/all.vs.{ident}.faa", ident=PHYLA),

rule clustalo:
    input:
        fasta = '../results/concatenated_files/metaT.Rdrphits.dedup.faa', 
        alignment = "../resources/RdRp-scan/Phylogenies/{ident}.CLUSTALO_0.4.fa",
    output:
        res='../results/alignments/all.vs.{ident}.faa',
    conda: 
        "clustalo"
    threads: 12
    shell:
        """
        clustalo --p1 {input.alignment} \
        --auto -i {input.fasta} -t Protein --threads={threads} \
        -o {output.res}
        """